# Capstone — mirrors your deployed research paper

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. Question

*The research question and the decision it supports.*

In [2]:
print("""
Research Question

Can a Random Forest Regression model accurately estimate Click-Through Rate (CTR) using production search features?

Decision Supported

The model is intended to support content ranking decisions by estimating CTR from historical search data. The results are used as decision-support evidence rather than proof of future performance.
""")


Research Question

Can a Random Forest Regression model accurately estimate Click-Through Rate (CTR) using production search features?

Decision Supported

The model is intended to support content ranking decisions by estimating CTR from historical search data. The results are used as decision-support evidence rather than proof of future performance.



## 2. Data

*Which release, which tables, date windows, what you excluded and why. Public-safe.*

In [3]:
print("""
Data

Dataset:
FlyRank ML Internship dataset (content_refresh_anonymized.csv)

Target:
CTR (Click-Through Rate)

Features:
Numerical production search features such as search volume, CPC, impressions, clicks, sessions, users, engagement metrics, and content age.

Date Window:
The dataset provided for the internship was used without modification.

Excluded Data:
Non-numeric features and rows with missing values were excluded to simplify model training and ensure data quality.
""")


Data

Dataset:
FlyRank ML Internship dataset (content_refresh_anonymized.csv)

Target:
CTR (Click-Through Rate)

Features:
Numerical production search features such as search volume, CPC, impressions, clicks, sessions, users, engagement metrics, and content age.

Date Window:
The dataset provided for the internship was used without modification.

Excluded Data:
Non-numeric features and rows with missing values were excluded to simplify model training and ensure data quality.



## 3. Methodology

*Assumptions, features, label definition, baseline, validation design, leakage checks.*

In [4]:
print("""
Methodology

Model:
RandomForestRegressor (100 trees, random_state=42)

Assumptions:
The available numerical features contain useful information for estimating CTR.

Label:
CTR

Baseline:
A simple baseline from previous work was used for comparison on the same train/test split.

Validation:
80/20 random train/test split.

Leakage Checks:
Potential leakage was reviewed in the validation audit. Future work should use GroupKFold or TimeSeriesSplit for a more realistic evaluation.
""")


Methodology

Model:
RandomForestRegressor (100 trees, random_state=42)

Assumptions:
The available numerical features contain useful information for estimating CTR.

Label:
CTR

Baseline:
A simple baseline from previous work was used for comparison on the same train/test split.

Validation:
80/20 random train/test split.

Leakage Checks:
Potential leakage was reviewed in the validation audit. Future work should use GroupKFold or TimeSeriesSplit for a more realistic evaluation.



## 4. Results (vs baseline)

*Model vs baseline on the same split. The honest table.*

In [5]:
print("""
Results

Model Performance

MAE = 0.015586
MSE = 0.016527
R² = 0.976225

Compared with the baseline, the Random Forest model achieved better predictive performance on the same evaluation split.

The measured results suggest that the model captures most of the variation in CTR within this dataset.
""")


Results

Model Performance

MAE = 0.015586
MSE = 0.016527
R² = 0.976225

Compared with the baseline, the Random Forest model achieved better predictive performance on the same evaluation split.

The measured results suggest that the model captures most of the variation in CTR within this dataset.



## 5. Limitations

*What this work cannot claim.*

In [6]:
print("""
Limitations

The model was evaluated only on the available internship dataset.

The random train/test split may produce optimistic estimates because related observations could appear in both training and testing sets.

The results should be interpreted as decision-support evidence rather than proof of generalization to future production data.
""")


Limitations

The model was evaluated only on the available internship dataset.

The random train/test split may produce optimistic estimates because related observations could appear in both training and testing sets.

The results should be interpreted as decision-support evidence rather than proof of generalization to future production data.



## 6. Ranked recommendations

*The action playbook output — the paper's recommendations section.*

In [7]:
print("""
Ranked Recommendations

1. Validate the model using GroupKFold.
2. Evaluate performance with TimeSeriesSplit.
3. Monitor model performance after deployment.
4. Retrain the model when new data becomes available.
5. Compare Random Forest with Gradient Boosting and XGBoost models.
""")


Ranked Recommendations

1. Validate the model using GroupKFold.
2. Evaluate performance with TimeSeriesSplit.
3. Monitor model performance after deployment.
4. Retrain the model when new data becomes available.
5. Compare Random Forest with Gradient Boosting and XGBoost models.



## 7. Artifacts the paper embeds

*Generate/collect the charts and tables your deployed page will show.*

In [8]:
print("""
Artifacts

The paper includes:

- Model evaluation metrics (MAE, MSE, R²).
- Model vs baseline comparison.
- Prediction error analysis.
- Leakage audit summary.
- Validation discussion.
- Ranked recommendations for future work.
""")


Artifacts

The paper includes:

- Model evaluation metrics (MAE, MSE, R²).
- Model vs baseline comparison.
- Prediction error analysis.
- Leakage audit summary.
- Validation discussion.
- Ranked recommendations for future work.



# Demo Outline (5 Minutes)

## 1. Problem
Predict Click-Through Rate (CTR) using production search features from the FlyRank dataset.

## 2. Method
I trained a Random Forest Regressor using numerical features after preprocessing the dataset.

## 3. One Chart
Model evaluation table showing MAE, MSE, and R².

## 4. Honest Result
The model achieved an R² score of 0.976 on the evaluation split. Additional grouped or time-aware validation is recommended.

## 5. Recommendation
Use GroupKFold or TimeSeriesSplit before production deployment and continue monitoring model performance.

# Social Post

I completed my FlyRank ML Internship capstone project.

I built a Random Forest Regression model to predict Click-Through Rate (CTR) using production search features.

The project focused on reproducible machine learning, honest validation, and leakage-aware evaluation.

# Employer Summary

I built a Random Forest Regression model using the FlyRank ML Internship dataset to predict Click-Through Rate (CTR).

The model achieved an R² score of 0.976 on the evaluation dataset.

The project emphasized reproducible machine learning, careful validation, and decision-support analysis.

In [9]:
!git add .
!git commit -m "Finalize capstone"
!git push

fatal: not a git repository (or any of the parent directories): .git
fatal: not a git repository (or any of the parent directories): .git
fatal: not a git repository (or any of the parent directories): .git


## Self-check

Before you submit, confirm each line honestly:

- [ ] Every section above is filled — markdown thinking AND the code that backs it
- [ ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [ ] No client names, URLs, or private queries anywhere
- [ ] My claims use careful words: observed, measured, directional, decision-support
- [ ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.